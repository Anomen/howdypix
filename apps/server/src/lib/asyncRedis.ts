import { promisify } from "util";
import redis, { ClientOpts } from "redis";

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export const asyncRedis = (options?: ClientOpts, redisClient = redis) => {
  const client = redisClient.createClient(options);

  const get = promisify(client.get).bind(client);
  const set = promisify(client.set).bind(client);
  const flushdb = promisify(client.flushdb).bind(client);

  const getJson = <T>(arg1: string): Promise<T | null> =>
    get(arg1).then((str) => (str ? (JSON.parse(str) as T) : null));
  const setJson = <T>(arg1: string, arg2: T): Promise<unknown> =>
    set(arg1, JSON.stringify(arg2));

  return {
    get,
    set,
    flushdb,
    getJson,
    setJson,
  };
};

export type AsyncRedis = ReturnType<typeof asyncRedis>;
