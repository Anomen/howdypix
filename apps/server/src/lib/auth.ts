import jwt from "jsonwebtoken";
import { appDebug } from "@howdypix/utils";
import { TokenInfo, UserInfo } from "@howdypix/shared-types";
import { User, UserConfigState } from "../state";
import config from "../config";

const debug = appDebug("lib:auth");

/**
 * The Stores is an object in the memory to store the current valid codes that
 * have been sent by email, and validate the refreshTokens to make sure they are valid.
 */
type Stores = {
  codes: {
    [email: string]: string;
  };
  tokenList: {
    [refreshToken: string]: TokenInfo;
  };
};

const stores: Stores = {
  codes: {},
  tokenList: {}
};

const trunkToken = (token: string): string => {
  if (token.length > 5) {
    return `${token.substr(0, 5)}...${token.substr(-5)}`;
  }

  return token;
};

// The whole point of predicate functions are to define the type, so we disable the eslint rule.
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const isUserInfo = (data: any): data is UserInfo => {
  return data && !!data.email && !!data.name;
};

//= ===================================================
// Validation functions
//= ===================================================
export const isEmailValid = (
  authorizedUsers: UserConfigState["users"],
  emailToCheck: string
): User | null => authorizedUsers.find(u => u.email === emailToCheck) || null;

export const isJwtTokenValid = async (
  token: string,
  secret: string
): Promise<UserInfo | null> =>
  new Promise<UserInfo | null>(resolve => {
    debug(`Token: ${trunkToken(token)}`);

    jwt.verify(token, secret, (error, decoded) => {
      if (!error && isUserInfo(decoded)) {
        debug(`Token valid!`, decoded);
        resolve({ email: decoded.email, name: decoded.name });
      } else {
        debug(`Token not valid!`);
        resolve(null);
      }
    });
  });

export const isTokenValid = (token: string): Promise<UserInfo | null> => {
  debug(`Check if the token is valid...`);
  return isJwtTokenValid(token, config.auth.token.secret);
};

export const isRefreshTokenValid = (
  token: string
): Promise<UserInfo | null> => {
  debug(`Check if the refresh token is valid...`);
  return isJwtTokenValid(token, config.auth.refreshToken.secret);
};

export const isCodeValid = async (code: string): Promise<UserInfo | null> => {
  debug(`Check if the user code is valid...`);
  const user = await isJwtTokenValid(code, config.auth.code.secret);
  if (user && stores.codes[user.email]) {
    return user;
  }
  return null;
};

//= ===================================================
// Generate functions
//= ===================================================
export const generateJwtToken = async (
  user: UserInfo,
  options: { secret: string; expiry: string }
): Promise<string> =>
  new Promise((resolve, reject) => {
    jwt.sign(
      { ...user },
      options.secret,
      { expiresIn: options.expiry },
      (err, token) => {
        if (err) {
          reject(err);
        } else {
          debug(`Token: ${trunkToken(token)}`);
          resolve(token);
        }
      }
    );
  });

export const generateCode = (user: UserInfo): Promise<string> => {
  debug("Generate the code for the magic link.", user);
  return generateJwtToken(user, config.auth.code);
};

export const generateToken = (user: UserInfo): Promise<string> => {
  debug("Generate the API token.", user);
  return generateJwtToken(user, config.auth.token);
};

export const generateRefreshToken = (user: UserInfo): Promise<string> => {
  debug("Generate the Refresh token.", user);
  return generateJwtToken(user, config.auth.refreshToken);
};

export const generateTokens = async (user: UserInfo): Promise<TokenInfo> =>
  new Promise(resolve => {
    debug("Generate the API tokens for auth.", user);

    generateToken(user).then(token => {
      generateRefreshToken(user).then(refreshToken => {
        resolve({ token, refreshToken, user });
      });
    });
  });

//= ===================================================
// Storage functions
//= ===================================================
export const storeCode = (email: string, code: string): void => {
  stores.codes[email] = code;
  debug("Saved code in memory:", stores.codes);
};

export const removeCode = (email: string): void => {
  delete stores.codes[email];
  debug("Removed code from memory:", email);
};

export const storeRefreshToken = (tokens: TokenInfo): void => {
  stores.tokenList[tokens.refreshToken] = tokens;
  debug("Saved refreshToken in memory:", tokens.refreshToken);
};

export const removeRefreshToken = (refreshToken: string): void => {
  delete stores.tokenList[refreshToken];
  debug("Removed refreshToken from memory:", refreshToken);
};
