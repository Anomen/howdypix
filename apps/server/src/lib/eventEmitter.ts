import StrictEventEmitter from "strict-event-emitter-types";
import { EventEmitter } from "events";
import { ProcessData } from "@howdypix/shared-types";

export interface EventTypes {
  newFile: { root: string; path: string; sourceId: string };
  removeFile: { root: string; path: string; sourceId: string };
  processFile: { root: string; path: string; sourceId: string };
  processedFile: ProcessData;
}

export type Events = StrictEventEmitter<EventEmitter, EventTypes>;
