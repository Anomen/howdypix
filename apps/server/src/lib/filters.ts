import { In } from "typeorm";
import { flatten } from "lodash";
import { AvailableFilters } from "@howdypix/shared-types";

//= =============================================================
// Define the types
//= =============================================================
type FilterValue = string | (string | null)[] | null | undefined;
type FilterFunctionReturn = {
  whereStatement: { [key: string]: string } | {};
};

type FilterFunction = (filterValue: FilterValue) => FilterFunctionReturn;

//= =============================================================
// Define the generic functions
//= =============================================================
export const filterWithText: (
  property: keyof AvailableFilters
) => FilterFunction = property => (value): FilterFunctionReturn => {
  return {
    whereStatement: value
      ? {
          [property]: In(flatten([value]))
        }
      : {}
  };
};

//= =============================================================
// Define the filter functions
//= =============================================================
export const filterByMake = filterWithText("make");
export const filterByModel = filterWithText("model");
