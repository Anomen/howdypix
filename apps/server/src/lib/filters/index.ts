import { filterByMake } from "./filterByMake";
import { FilterValue, FilterFunctionReturn } from "./filterTypes";

export function filterFactory(
  filterType: string,
  filterValue: string | string[]
): FilterFunctionReturn | null {
  if (filterType === "make") {
    return filterByMake(filterValue);
  }

  return null;
}
