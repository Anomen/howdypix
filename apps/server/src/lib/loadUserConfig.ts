import fs from "fs";
import { join } from "path";
import { parse } from "yaml";
import { state } from "../state";

export function loadUserConfig() {
  const content = fs.readFileSync(join(__dirname, "..", "..", "config.yaml"));

  state.userConfig = parse(content.toString());
  state.userConfig.thumbnailsDir = join(
    process.cwd(),
    state.userConfig.thumbnailsDir
  );

  return state.userConfig;
}
