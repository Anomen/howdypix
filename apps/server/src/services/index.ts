export { startRabbitMq } from "./startRabbitMq";
export { startCacheDB } from "./startCacheDB";
export { startFileScan } from "./startFileScan";
