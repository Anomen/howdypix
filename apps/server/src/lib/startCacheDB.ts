import { Connection } from "typeorm";
import { appDebug, generateThumbnailPaths, hjoin } from "@howdypix/utils";
import { existsSync, statSync, unlinkSync } from "fs";
import { join, parse } from "path";
import { UserConfig } from "../config";
import { Photo } from "../entity/Photo";
import { Events, EventTypes } from "./eventEmitter";
import { Source } from "../entity/Source";
import { Album } from "../entity/Album";

export async function onNewDir(
  { root, hfile }: EventTypes["newDirectory"],
  event: Events,
  connection: Connection
): Promise<void> {
  const absolutePath = join(root, hfile.dir ?? "", hfile.file ?? "");
  const stat = statSync(absolutePath);

  // Check if it exists in the database
  const albumRepository = connection.getRepository(Album);
  const sourceRepository = connection.getRepository(Source);
  const album = await albumRepository.findOne({ where: { inode: stat.ino } });

  // Check the updated_date
  if (!album) {
    const newAlbum = new Album();
    newAlbum.dir = hfile.file ?? "";
    newAlbum.parentDir = hfile.dir ?? "";
    newAlbum.source = hfile.source;
    newAlbum.inode = stat.ino.toString();

    newAlbum.sourceLk =
      (await sourceRepository.findOne({
        source: hfile.source,
      })) ?? null;

    albumRepository.save(newAlbum);
  }
}

export async function onUnlinkDir(
  { root, hfile }: EventTypes["unlinkDirectory"],
  event: Events,
  connection: Connection
): Promise<void> {
  const absolutePath = join(root, hfile.dir ?? "", hfile.file ?? "");
  const stat = statSync(absolutePath);

  // Check if it exists in the database
  const albumRepository = connection.getRepository(Album);
  await albumRepository.delete({ inode: stat.ino.toString() });
}

export async function onNewFile(
  { root, hfile }: EventTypes["newFile"],
  event: Events,
  connection: Connection,
  userConfig: UserConfig
): Promise<void> {
  const absolutePath = join(root, hfile.dir ?? "", hfile.file ?? "");
  const stat = statSync(absolutePath);

  // Check if it exists in the database
  const photoRepository = connection.getRepository(Photo);
  const photo = await photoRepository.findOne({ inode: stat.ino });

  // Check the updated_date
  if (photo?.mtime === stat.mtimeMs) {
    const thumbnailsExist: boolean = generateThumbnailPaths(
      userConfig.thumbnailsDir,
      hfile
    ).reduce(
      (accumulator: boolean, tn) => accumulator && existsSync(tn.path),
      true
    );

    if (thumbnailsExist) {
      return;
    }
  }

  // Act
  event.emit("processFile", { root, hfile });
}

export async function onRemoveFile(
  { hfile }: EventTypes["newFile"],
  event: Events,
  connection: Connection,
  userConfig: UserConfig
): Promise<void> {
  const photoRepository = connection.getRepository(Photo);

  // Remove from database
  await photoRepository.delete(hfile);

  // Remove thumbnails
  generateThumbnailPaths(userConfig.thumbnailsDir, hfile).forEach((tn) => {
    unlinkSync(tn.path);
  });
}

export async function onProcessedFile(
  file: EventTypes["processedFile"],
  event: Events,
  connection: Connection
): Promise<void> {
  const albumRepository = connection.getRepository(Album);
  const photoRepository = connection.getRepository(Photo);

  const photo = new Photo();
  photo.make = file.exif.make ?? "";
  photo.model = file.exif.model ?? "";
  photo.ISO = file.exif.ISO ?? 0;
  photo.shutter = file.exif.shutter ?? 0;
  photo.processedShutter = file.exif.shutter
    ? Math.round((1 / file.exif.shutter) * 10) / 10
    : 0;
  photo.aperture = file.exif.aperture ?? 0;
  photo.processedAperture = file.exif.aperture
    ? Math.round(file.exif.aperture * 10) / 10
    : 0;
  photo.createDate = file.exif.createDate ?? 0;
  photo.inode = file.stat.inode;
  photo.mtime = file.stat.mtime;
  photo.ctime = file.stat.ctime;
  photo.birthtime = file.stat.birthtime;
  photo.size = file.stat.size;
  photo.source = file.hfile.source;
  photo.parentDir = parse(file.hfile.dir ?? "").dir ?? "";
  photo.dir = file.hfile.dir ?? "";
  photo.file = file.hfile.file ?? "";

  photo.album =
    (await albumRepository.findOne({
      dir: file.hfile.dir ?? "",
      source: file.hfile.source,
      parentDir: parse(file.hfile.dir ?? "").dir ?? "",
    })) ?? null;

  await photoRepository.save(photo);
  appDebug("db")(`Saved ${hjoin(file.hfile)}.`);
}

export async function startCacheDB(
  event: Events,
  userConfig: UserConfig,
  connection: Connection
): Promise<Connection> {
  // Insert the source entries in the database
  await Promise.all(
    Object.keys(userConfig.photoDirs).map(async (source) => {
      await Source.upsert(source, userConfig.photoDirs[source]);
    })
  );

  event.on("newFile", (params) =>
    onNewFile(params, event, connection, userConfig)
  );

  event.on("removeFile", (params) =>
    onRemoveFile(params, event, connection, userConfig)
  );

  event.on("newDirectory", (params) => onNewDir(params, event, connection));

  event.on("unlinkDirectory", (params) =>
    onUnlinkDir(params, event, connection)
  );

  event.on("processedFile", (file) => onProcessedFile(file, event, connection));

  return connection;
}
