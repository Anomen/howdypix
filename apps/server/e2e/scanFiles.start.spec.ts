import { statSync } from "fs";
import { appError } from "@howdypix/utils";
import { startCacheDB } from "../src/services/startCacheDB";
import { Events } from "../src/lib/eventEmitter";
import { initialize } from "./scanFiles.setup";

jest.mock("@howdypix/utils", () => ({
  ...jest.requireActual("@howdypix/utils"),
  appError: jest.fn(),
}));

jest.mock("fs", (): { [name: string]: jest.Mock } => ({
  ...jest.requireActual("fs"),
  statSync: jest.fn(),
}));

describe("[Start] Scanning files", (): void => {
  let events: Events;

  const {
    pathHashCode,
    baseUserConfig,
    assertDatabaseValue,
    resetScanFilesTests,
    connection,
  } = initialize();

  // ===================================================
  // Reset after each tests
  // ===================================================
  beforeEach(
    async (): Promise<void> => {
      jest.resetAllMocks();
      jest.restoreAllMocks();

      const setup = await resetScanFilesTests();
      events = setup.events;
    }
  );

  // ===================================================
  // Tests
  // ===================================================
  describe("when starting the process", (): void => {
    test("should add a new source entry", async (): Promise<void> => {
      await startCacheDB(events, baseUserConfig, connection);

      // Test if the album entries have been created
      assertDatabaseValue();
    });

    test("should display an error if one source does not have the directory accessible but the other yes", async (): Promise<
      void
    > => {
      (statSync as jest.Mock).mockImplementation((name) => {
        if (/first$/.exec(name)) {
          throw new Error("ENOENT: no such file or directory, stat '/first'");
        } else {
          return {
            ino: pathHashCode(name),
          };
        }
      });

      await startCacheDB(events, baseUserConfig, connection);

      // Test if the album entries have been created
      expect(appError).toHaveBeenCalled();
      assertDatabaseValue();
    });
  });
});
