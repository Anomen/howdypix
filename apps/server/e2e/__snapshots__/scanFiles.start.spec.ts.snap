// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Start] Scanning files when starting the process should display an error if one source does not have the directory accessible but the other yes 1`] = `
Array [
  Object {
    "dir": "",
    "id": 1,
    "inode": "c2VydmVyZmlyc3Q=",
    "parentDir": "",
    "photos": Array [],
    "source": "first",
    "sourceLk": Source {
      "albums": undefined,
      "dir": "./first",
      "id": 2,
      "source": "first",
    },
  },
  Object {
    "dir": "",
    "id": 1,
    "inode": "c2VydmVyc2Vjb25k",
    "parentDir": "",
    "photos": Array [],
    "source": "second",
    "sourceLk": Source {
      "albums": undefined,
      "dir": "./second",
      "id": 1,
      "source": "second",
    },
  },
]
`;

exports[`[Start] Scanning files when starting the process should display an error if one source does not have the directory accessible but the other yes 2`] = `
Array [
  Object {
    "albums": Array [
      Album {
        "dir": "",
        "id": 2,
        "inode": "c2VydmVyZmlyc3Q=",
        "parentDir": "",
        "photos": undefined,
        "source": "first",
        "sourceLk": undefined,
      },
    ],
    "dir": "./first",
    "id": 1,
    "source": "first",
  },
  Object {
    "albums": Array [
      Album {
        "dir": "",
        "id": 1,
        "inode": "c2VydmVyc2Vjb25k",
        "parentDir": "",
        "photos": undefined,
        "source": "second",
        "sourceLk": undefined,
      },
    ],
    "dir": "./second",
    "id": 1,
    "source": "second",
  },
]
`;
